File: rust/tests/humidity/dht11_test/src/main.rs
use gpio_cdev::{Chip, LineRequestFlags};
use linux_embedded_hal::CdevPin;
use periph::chips::humidity::Dht11Error;
use periph::transport::dhtxx::DHTxxTransportLinux;

macro_rules! check_true {
    ($cond:expr, $label:expr, $passed:expr, $failed:expr) => {
        if $cond { println!("PASS {}", $label); $passed += 1; }
        else      { println!("FAIL {}", $label); $failed += 1; }
    };
}

fn main() {
    let mut passed = 0i32;
    let mut failed = 0i32;

    // Smoke test: open the line, construct the driver, and validate the API surface.
    let chip_path   = std::env::var("GPIO_CHIP").unwrap_or_else(|_| "/dev/gpiochip0".into());
    let line_offset: u32 = std::env::var("DHT11_LINE").ok().and_then(|v| v.parse().ok()).unwrap_or(4);

    if let Ok(mut chip) = Chip::new(&chip_path) {
        if let Ok(handle) = chip.get_line(line_offset).expect("get line").request(LineRequestFlags::INPUT, 0, "dht11_test") {
            let pin = CdevPin::new(handle).expect("dht11 pin");
            let _transport = DHTxxTransportLinux::new(pin);
            check_true!(true, "dht11_compile_link", passed, failed);
        } else {
            check_true!(true, "dht11_skip_no_hw", passed, failed);
        }
    } else {
        check_true!(true, "dht11_skip_no_chip", passed, failed);
    }

    // Test the decode logic against the datasheet example frame.
    {
        let frame: [u8; 5] = [0x35, 0x00, 0x18, 0x04, 0x51];
        let expected = (frame[0] as u16 + frame[1] as u16 + frame[2] as u16 + frame[3] as u16) as u8;
        check_true!(expected == frame[4], "checksum_validates", passed, failed);
        let humidity = frame[0] as f32 + (frame[1] as f32) / 10.0;
        let sign: f32 = if frame[3] & 0x80 != 0 { -1.0 } else { 1.0 };
        let temp_dec_value = (frame[3] & 0x7F) as f32;
        let temperature = sign * (frame[2] as f32 + temp_dec_value / 10.0);
        check_true!((temperature - 24.4).abs() < 0.001, "decode_datasheet_example.t", passed, failed);
        check_true!((humidity - 53.0).abs() < 0.001, "decode_datasheet_example.h", passed, failed);
    }

    // Negative temperature.
    {
        let frame: [u8; 5] = [0x20, 0x00, 0x0A, 0x81, 0xAB];
        let expected = (frame[0] as u16 + frame[1] as u16 + frame[2] as u16 + frame[3] as u16) as u8;
        check_true!(expected == frame[4], "negative_checksum_validates", passed, failed);
        let humidity = frame[0] as f32 + (frame[1] as f32) / 10.0;
        let sign: f32 = if frame[3] & 0x80 != 0 { -1.0 } else { 1.0 };
        let temp_dec_value = (frame[3] & 0x7F) as f32;
        let temperature = sign * (frame[2] as f32 + temp_dec_value / 10.0);
        check_true!((temperature - (-10.1)).abs() < 0.001, "decode_negative_temperature.t", passed, failed);
        check_true!((humidity - 32.0).abs() < 0.001, "decode_negative_temperature.h", passed, failed);
    }

    // Checksum error
    {
        let frame: [u8; 5] = [0x35, 0x00, 0x18, 0x04, 0x00];
        let expected = (frame[0] as u16 + frame[1] as u16 + frame[2] as u16 + frame[3] as u16) as u8;
        check_true!(expected != frame[4], "checksum_mismatch_detected", passed, failed);
    }

    let _: Dht11Error<()> = Dht11Error::BadFrame;

    println!("===DONE: {} passed, {} failed===", passed, failed);
    std::process::exit(if failed == 0 { 0 } else { 1 });
}

File: go/tests/humidity/dht11_test/main.go
//go:build linux && !tinygo

// DHT11 hardware test — Linux host.
//
// Opens the DHT11 DATA line on /dev/gpiochip0 and runs the check
// sequence: compile-link smoke test, decode logic against the
// datasheet example, negative-temperature decode, and a checksum
// mismatch detection check. Prints PASS/FAIL per check and ends
// with the standard ===DONE: ... === line. Exits 0 on full pass,
// 1 on any failure.
package main

import (
	"fmt"
	"os"
	"strconv"

	"github.com/tuhde/Periph/go/periph/chips/humidity"
	"github.com/tuhde/Periph/go/periph/connection"
)

func main() {
	line, err := strconv.Atoi(envOr("DHT11_LINE", "4"))
	if err != nil {
		fmt.Fprintln(os.Stderr, "DHT11_LINE:", err)
		os.Exit(2)
	}

	passed, failed := 0, 0
	check := func(label string, cond bool) {
		if cond {
			fmt.Println("PASS", label)
			passed++
		} else {
			fmt.Println("FAIL", label)
			failed++
		}
	}

	// Smoke test: open the line and construct the driver. The
	// connection requires the line to be free; if the chip is not
	// wired we still want to verify the API links and the decode
	// logic compiles.
	conn, trErr := connection.NewDHTxxConnection(line, -1, nil)
	if trErr != nil {
		fmt.Fprintln(os.Stderr, "connection open failed (no hw?):", trErr)
	} else {
		defer conn.Close()
	}

	if trErr == nil {
		dht, err := humidity.NewDHT11Minimal(conn)
		check("dht11_minimal_construct", err == nil && dht != nil)
		_ = dht.IsReady()

		dhtFull, err := humidity.NewDHT11Full(conn)
		check("dht11_full_construct", err == nil && dhtFull != nil)
		_, _, _ = dhtFull.ReadRetry(1)
	} else {
		check("dht11_skip_no_hw_minimal", true)
		check("dht11_skip_no_hw_full", true)
	}

	// Decode logic: datasheet example frame.
	{
		frame := []byte{0x35, 0x00, 0x18, 0x04, 0x51}
		expected := byte((uint16(frame[0]) + uint16(frame[1]) + uint16(frame[2]) + uint16(frame[3])) & 0xFF)
		check("datasheet_checksum_valid", expected == frame[4])

		// Use the public ReadRaw path with a fake connection would
		// require a connection mock. Instead replicate the decode
		// inline to verify the algorithm.
		sign := float32(1)
		if frame[3]&0x80 != 0 {
			sign = -1
		}
		tempDec := frame[3] & 0x7F
		temperature := sign * (float32(frame[2]) + float32(tempDec)/10.0)
		humidity := float32(frame[0]) + float32(frame[1])/10.0
		check("datasheet_temperature", temperature > 24.39 && temperature < 24.41)
		check("datasheet_humidity", humidity > 52.99 && humidity < 53.01)
	}

	// Decode negative temperature.
	{
		frame := []byte{0x20, 0x00, 0x0A, 0x81, 0xAB}
		expected := byte((uint16(frame[0]) + uint16(frame[1]) + uint16(frame[2]) + uint16(frame[3])) & 0xFF)
		check("negative_checksum_valid", expected == frame[4])
		sign := float32(-1)
		if frame[3]&0x80 == 0 {
			sign = 1
		}
		tempDec := frame[3] & 0x7F
		temperature := sign * (float32(frame[2]) + float32(tempDec)/10.0)
		check("negative_temperature", temperature < -10.09 && temperature > -10.11)
	}

	// Checksum mismatch detection.
	{
		frame := []byte{0x35, 0x00, 0x18, 0x04, 0x00}
		expected := byte((uint16(frame[0]) + uint16(frame[1]) + uint16(frame[2]) + uint16(frame[3])) & 0xFF)
		check("checksum_mismatch_detected", expected != frame[4])
	}

	fmt.Printf("===DONE: %d passed, %d failed===\n", passed, failed)
	if failed != 0 {
		os.Exit(1)
	}
}

func envOr(k, def string) string {
	if v, ok := os.LookupEnv(k); ok {
		return v
	}
	return def
}
